### PR TITLE
Limit the number of spec files downloaded to find matches for buildcaches

### PR DESCRIPTION
**Pull request by gartung**
Further limit the number of spec files downloaded by spack install process when it looks for buildcaches. Use the package name or hash in a regex along with current platform and os to match urls for spec files before they are downloaded.

### CODE REVIEW

**Comment by gartung** [comment]
@scottwittenburg would you be willing to test this is your build pipeline?

**Comment by scottwittenburg** [comment]
@gartung Sure I can do that.  And just so I understand, the goal is to always find the binary we need on the mirror, but without having to download every `spec.yaml`, is that right?

**Comment by gartung** [comment]
Yes. I saves downloading all of the specs.

**Comment by gartung** [comment]
Looks like I could use some advice on why the cmd/ci test is failing.

**Comment by scottwittenburg** [comment]
I can try to take a look at that failing test tomorrow.

**Comment by gartung** [comment]
@scottwittenburg I made changes that fixed the test. I just need someone to verify that this PR doesn't break anything. 
@eugeneswalker you might be interested in testing this PR too.

**Comment by scottwittenburg** [comment]
Sorry today and tomorrow I'm quite busy on another project, so I may not get a chance to try until Friday.

**Comment by eugeneswalker** [comment]
@gartung this only affects reading from cache during `spack install` ? is that the intention here? With this PR I still see all OS/Arch combinations in `spack buildcache list` -- is that expected?

**Comment by gartung** [comment]
I also added the ability to filter by the current arch with the -a (--arch) flag.In addition the spec files downloaded are filtered by the spec names added.

**Comment by gartung** [comment]
So you can do
spack buildcache list -a
spack buildcache list python zlib
and it will only download spec that match the arch or match a regex of the names.

File: lib/spack/spack/binary_distribution.py [hunk 1]
```diff
@@ -664,14 +664,23 @@ def extract_tarball(spec, filename, allow_root=False, unsigned=False,
 _cached_specs = None
 
 
-def get_specs(force=False, use_arch=False):
+def get_specs(force=False, use_arch=False, names=[]):
```
Comment (alalazo) [inline lib/spack/spack/binary_distribution.py]: @gartung Python gotcha: [don't use mutable arguments as default parameters](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments)

**Comment by gartung** [comment]
@alalazo see https://github.com/spack/spack/pull/14696

**Comment by scottwittenburg** [comment]
@gartung: I finally got a chance to try it out today, as I mentioned I would.  It seems like it's filtering a little too aggressively, because a matching dag hash in the binary mirror is not found, breaking my most recent pipeline (because I install dependencies with `--cache-only`).  Did this come with a way to turn the option off?

**Comment by gartung** [comment]
Can you try the head of develop.

**Comment by scottwittenburg** [comment]
Yes, that's what I just tried.

**Comment by gartung** [comment]
So spack install --cache-only /spec ?

**Comment by gartung** [comment]
See https://github.com/spack/spack/pull/14698 we can carry on the discussion there.

**Comment by scottwittenburg** [comment]
No, it's for the installation of dependencies: `spack install --only dependencies --cache-only <spec>`

**Comment by eugeneswalker** [comment]
same for me. i'm installing the same way and it is not finding packages as of HEAD of develop this morning

**Comment by gartung** [comment]
Trying to reproduce ther problem locally. Maybe I can fix it once I can reproduce the error.

**Comment by gartung** [comment]
https://github.com/spack/spack/pull/14714 fixed 
spack install --only package --no-cache
in my local testing.

**Comment by gartung** [comment]
I needed to fix how the spec caching was done,

**Pull request merged**